File: 092__Reverse_Linked_List_II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """
        # Consider linked list as a list A. i.e. A[1]=head.val, A[2]=head.next.val ....
        # Create dummy.         [D,A[1],A[2],....]
        dummy = ListNode(None)
        dummy.next = head
        # prev at A[0]
        prev = dummy

        # move prev m-1 times. now prev is at A[M-1]
        for _ in range(m -1):
            prev = prev.next

        # start reversing from curr = prev.next (A[M]) for n-m+1 nodes
        curr, temp = prev.next, None
        for _ in range(n - m + 1):
            curr.next, curr, temp = temp, curr.next, curr
            # Or more clear:
            # temp = curr.next
            # curr.next = node
            # node = curr
            # curr = temp


        # now the situation is
        # [dummy .... prev]    [temp,..., prev.next]   [curr .....]
        # connect prev to temp, and prev.next to curr
        prev.next.next, prev.next = curr, temp

        return dummy.next

File: Java/092__Reverse_Linked_List_II.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {

        if (head == null) {
            return null;
        }
        ListNode rest = head, prev = null;
        while (m > 1) {
            prev = rest;
            rest = rest.next;
            m--;
            n--;
        }

        ListNode unchanged_end = prev, changed_end = rest;

        // Iteratively reverse the nodes until n becomes 0.
        ListNode temp = null;
        while (n > 0) {
            temp = rest.next;
            rest.next = prev;
            prev = rest;
            rest = temp;
            n--;
        }

        if (unchanged_end != null) {
            unchanged_end.next = prev;
        } else {
            head = prev;
        }

        changed_end.next = rest;
        return head;
    }
}
